### ISSUE TITLE
ZMQServer "Operation cannot be accomplished in current state" exception causes loop to get stalled

### ISSUE THREAD

**Issue by kinow** [cylc/cylc-flow#3414]
**Describe the bug**

Running `cylc run --no-detach --verbose --debug five` with Cylc 8/`master`, and then running `cylc gui five` with Cylc 7.8.x breaks the Cylc 8 workflow.

The traceback when that happens is the following:

```bash
2019-10-21T05:56:28Z DEBUG - 0 task(s) de-queued
2019-10-21T05:56:28Z DEBUG - [prep.20130808T0000Z] -forced
	spawning
2019-10-21T05:56:28Z DEBUG - END TASK PROCESSING (took
	0.001538991928100586 seconds)
2019-10-21T05:56:38Z ERROR - failed to decode message: Not enough segments
	Traceback (most recent call last):
	  File "/home/kinow/Development/python/workspace/cylc-
	uiserver/venv/lib/python3.7/site-packages/jose/jws.py", line 181, in _load
	    signing_input, crypto_segment = jwt.rsplit(b'.', 1)
	ValueError: not enough values to unpack (expected 2, got 1)
	During handling of the above exception, another exception occurred:
	Traceback (most recent call last):
	  File "/home/kinow/Development/python/workspace/cylc-
	uiserver/venv/lib/python3.7/site-packages/jose/jwt.py", line 135, in decode
	    payload = jws.verify(token, key, algorithms, verify=verify_signature)
	  File "/home/kinow/Development/python/workspace/cylc-
	uiserver/venv/lib/python3.7/site-packages/jose/jws.py", line 74, in verify
	    header, payload, signing_input, signature = _load(token)
	  File "/home/kinow/Development/python/workspace/cylc-
	uiserver/venv/lib/python3.7/site-packages/jose/jws.py", line 185, in _load
	    raise JWSError('Not enough segments')
	jose.exceptions.JWSError: Not enough segments
	During handling of the above exception, another exception occurred:
	Traceback (most recent call last):
	  File "/home/kinow/Development/python/workspace/cylc-
	flow/cylc/flow/network/server.py", line 165, in _listener
	    message = self.decode(msg, self.secret())
	  File "/home/kinow/Development/python/workspace/cylc-
	flow/cylc/flow/network/authentication.py", line 54, in decrypt
	    message = jwt.decode(message, secret, algorithms=[HASH])
	  File "/home/kinow/Development/python/workspace/cylc-
	uiserver/venv/lib/python3.7/site-packages/jose/jwt.py", line 137, in decode
	    raise JWTError(e)
	jose.exceptions.JWTError: Not enough segments
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/network/server.py", line 156, in _listener
    msg = self.socket.recv_string()
  File "/home/kinow/Development/python/workspace/cylc-uiserver/venv/lib/python3.7/site-packages/zmq/sugar/socket.py", line 587, in recv_string
    msg = self.recv(flags=flags)
  File "zmq/backend/cython/socket.pyx", line 796, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 832, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 191, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/socket.pyx", line 186, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/checkrc.pxd", line 25, in zmq.backend.cython.checkrc._check_rc
zmq.error.ZMQError: Operation cannot be accomplished in current state
```

Then `CTRL+C` stops the workflow. Trying `cylc run --no-detach --verbose --debug five` worked this afternoon at NIWA. But it failed when I tried again now at home.

```bash
...
2019-10-21T06:03:23Z DEBUG - END TASK PROCESSING (took
	0.0012841224670410156 seconds)
2019-10-21T06:03:54Z ERROR - failed to decode message: Not enough segments
	Traceback (most recent call last):
	  File "/home/kinow/Development/python/workspace/cylc-
	uiserver/venv/lib/python3.7/site-packages/jose/jws.py", line 181, in _load
	    signing_input, crypto_segment = jwt.rsplit(b'.', 1)
	ValueError: not enough values to unpack (expected 2, got 1)
	During handling of the above exception, another exception occurred:
	Traceback (most recent call last):
	  File "/home/kinow/Development/python/workspace/cylc-
	uiserver/venv/lib/python3.7/site-packages/jose/jwt.py", line 135, in decode
	    payload = jws.verify(token, key, algorithms, verify=verify_signature)
	  File "/home/kinow/Development/python/workspace/cylc-
	uiserver/venv/lib/python3.7/site-packages/jose/jws.py", line 74, in verify
	    header, payload, signing_input, signature = _load(token)
	  File "/home/kinow/Development/python/workspace/cylc-
	uiserver/venv/lib/python3.7/site-packages/jose/jws.py", line 185, in _load
	    raise JWSError('Not enough segments')
	jose.exceptions.JWSError: Not enough segments
	During handling of the above exception, another exception occurred:
	Traceback (most recent call last):
	  File "/home/kinow/Development/python/workspace/cylc-
	flow/cylc/flow/network/server.py", line 165, in _listener
	    message = self.decode(msg, self.secret())
	  File "/home/kinow/Development/python/workspace/cylc-
	flow/cylc/flow/network/authentication.py", line 54, in decrypt
	    message = jwt.decode(message, secret, algorithms=[HASH])
	  File "/home/kinow/Development/python/workspace/cylc-
	uiserver/venv/lib/python3.7/site-packages/jose/jwt.py", line 137, in decode
	    raise JWTError(e)
	jose.exceptions.JWTError: Not enough segments
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/home/kinow/Development/python/anaconda3/lib/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/home/kinow/Development/python/workspace/cylc-flow/cylc/flow/network/server.py", line 156, in _listener
    msg = self.socket.recv_string()
  File "/home/kinow/Development/python/workspace/cylc-uiserver/venv/lib/python3.7/site-packages/zmq/sugar/socket.py", line 587, in recv_string
    msg = self.recv(flags=flags)
  File "zmq/backend/cython/socket.pyx", line 796, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 832, in zmq.backend.cython.socket.Socket.recv
  File "zmq/backend/cython/socket.pyx", line 191, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/socket.pyx", line 186, in zmq.backend.cython.socket._recv_copy
  File "zmq/backend/cython/checkrc.pxd", line 25, in zmq.backend.cython.checkrc._check_rc
zmq.error.ZMQError: Operation cannot be accomplished in current state

2019-10-21T06:13:22Z DEBUG - Performing suite health check
2019-10-21T06:13:22Z DEBUG - Loading site/user config files
```

**Release version(s) and/or repository branch(es) affected?**

`master`

**Steps to reproduce the bug**

**Expected behavior**

The Cylc 7 is broken, while the Cylc 8 server keeps running without issues.

**Screenshots**

![image](https://user-images.githubusercontent.com/304786/67182124-f9ef4980-f43a-11e9-8fba-e8416b093a12.png)

**Additional context**

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

**Comment by kinow** [comment]
Users probably won't use this. I was alternating the same suite between Cylc 7 & Cylc 8 to compare the output with families, while reading the code.

And at one point saw the traceback, when I forgot to stop the suite in Cylc 8. Probably this won't happen once we have auth implemented I think, but thought it was still worth reporting as possible bug?.

**Comment by hjoliver** [comment]
Doing this definitely should not bring down the Cylc 8 workflow.  Seems like an easy one to fix (`not enough values to unpack (expected 2, got 1)`). But yeah I guess the upcoming auth changes will make this bit of code irrelevant?

**Comment by kinow** [comment]
Yup, happy to close as not an issue if you think we should do it now? :+1:

**Comment by kinow** [comment]
Found out what was causing my suite to not initialize. The `~/bin/cylc` script I had copied to my `$PATH`!

For the record, here's my `$PATH`, after activating my virtualenvironment: `/home/kinow/Development/python/workspace/cylc-uiserver/venv/bin:/home/kinow/.nvm/versions/node/v12.8.0/bin:/home/kinow/bin:/opt/ant/bin::/home/kinow/bin:/home/kinow/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/usr/lib/jvm/java-8-oracle/bin:/usr/lib/jvm/java-8-oracle/db/bin:/usr/lib/jvm/java-8-oracle/jre/bin`

If I remove the `cylc` script from my `/home/kinow/bin/` directory, `cylc run --... five` works... will investigate more tomorrow, but for now I suspect there is some cylc subcommand, or maybe the job.sh, ignoring my `$PATH` value.

**Comment by kinow** [comment]
Testing with the NIWA laptop again, where I didn't see the same behaviour as with my home laptop. In both I'm using an editable pip installation of the latest master of `cylc-flow`.

The issue doesn't happen 100%. Sometimes I am able to open Cylc 7 GUI, which doesn't show anything, but I can still switch between view types without causing the exception in Cylc 8 suite.

So I've attached the debugger this morning, without any breakpoint, but with the configuration to stop on any exception.

![image](https://user-images.githubusercontent.com/304786/67243706-b3d6cc00-f4b4-11e9-8961-a1e5de93d80d.png)

Here's when the exception happened (note that the `msg` value is some strange character?)

![Screenshot_2019-10-22_10-28-51](https://user-images.githubusercontent.com/304786/67244709-cf42d680-f4b6-11e9-8026-a27b70fcdb61.png)

**Comment by kinow** [comment]
We are seeing an exception logged here, but the reason for the workflow server program stopping, is actually an uncaught exception in the ZMQServer main thread.

Reading the - excellent - docstring for `self.socket.recv_string`, it also includes the possible exceptions:

![Screenshot_2019-10-22_10-41-26](https://user-images.githubusercontent.com/304786/67245471-83912c80-f4b8-11e9-8a67-f304c3f554aa.png)

So it may raise a `ZMQError`, but we are catching only `zmq.error.Again`. To confirm this, I've added an extra `except zmq.error.ZMQError`, and now finally have the root cause for the bug: an `ZMQError` with **Operation cannot be accomplished in current state**.

Possibly an invalid sequence for REQ/REP? Anyway, once this happens, the application stops on that breakpoint for every iteration of the `while True` loop. Meaning that the socket is now... dead? Or in an invalid state?

![Screenshot_2019-10-22_10-45-05](https://user-images.githubusercontent.com/304786/67245664-05815580-f4b9-11e9-95c7-0a33d6fc24b7.png)

**Comment by kinow** [comment]
Must be related to the other error with same exception we got in Cylc UI Server: https://github.com/cylc/cylc-uiserver/issues/65

I guess it will be fixed if/when we move to a different ZMQ communication pattern, like PUB/SUB.

**Comment by dwsutherland** [comment]
> Must be related to the other error with same exception we got in Cylc UI Server: [cylc/cylc-uiserver#65](https://github.com/cylc/cylc-uiserver/issues/65)
> 
> I guess it will be fixed if/when we move to a different ZMQ communication pattern, like PUB/SUB.

Yes having the data-store sync in will help.

The error happens when the same REQ socket sends two requests without receiving a reply (or the REP socket sends two responses in a row)... as documented here:
https://learning-0mq-with-pyzmq.readthedocs.io/en/latest/pyzmq/patterns/client_server.html

Of course you can have multiple clients (and possibly sockets?) to avoid this.. But it's really a fault on our interim UIS design (which is about to change)

**Comment by hjoliver** [comment]
Is this still a problem?

**Comment by kinow** [comment]
Yes. I think it was going to be solved after some future work in the UI Server. I think it was not done yet (was it Pub/Sub?).

To simulate the issue, you can use the current UI. Just send release/hold several times and eventually the workflow crashes.

**Comment by dwsutherland** [comment]
> Yes. I think it was going to be solved after some future work in the UI Server. I think it was not done yet (was it Pub/Sub?).
> 
> To simulate the issue, you can use the current UI. Just send release/hold several times and eventually the workflow crashes.

We would need to Queue the mutations if we use REQ/REP ZeroMQ pattern, as you cannot hit send twice before receiving the first..

The other option is to change REQ/REP pattern to ROUTER/DEALER (or some other)..

**Comment by oliver-sanders** [comment]
Tried to replicate by calling hammering the workflow with play/pause commands, can't reproduce the crash.

@kinow are you able to replicate on master.

**Comment by kinow** [comment]
Cannot reproduce it now @oliver-sanders :tada: 

![image](https://user-images.githubusercontent.com/304786/131747133-37ba67e7-956b-4f79-bfc1-b8ec9817e7b6.png)

Cylc 8 workflow keeps running smoothly with no errors.
